core/async-derive/src/lib.rs

```diff
@@ -0,0 +1,418 @@
+                }
+                _ => {
+                    continue;
+                }
+            };
+            if last_segment.ident == "Sender" {
+                let message_type = arguments[0].clone();
+                message_types.push(quote!(#message_type));
+                input_types.push(quote!(#message_type));
+                input_extractors.push(quote!(msg));
+            } else if last_segment.ident == "AsyncSender" {
+                let message_type = arguments[0].clone();
+                let result_type = arguments[1].clone();
+                message_types.push(quote!(near_async::messaging::MessageExpectingResponse<#message_type, #result_type>));
+                input_types.push(quote!(#message_type));
+                input_extractors.push(quote!(msg.message));
+            }
+        }
+    }
+
+    let mut message_derives = proc_macro2::TokenStream::new();
+    let mut input_derives = proc_macro2::TokenStream::new();
+    for attr in ast.attrs {
+        if attr.path().is_ident("multi_send_message_derive") {
+            let Meta::List(metalist) = attr.meta else {
+                panic!("multi_send_message_derive must be a list");
+            };
+            message_derives = metalist.tokens;
+        } else if attr.path().is_ident("multi_send_input_derive") {
+            let Meta::List(metalist) = attr.meta else {
+                panic!("multi_send_input_derive must be a list");
+            };
+            input_derives = metalist.tokens;
+        }
+    }
+
+    quote! {
+        #[derive(#message_derives)]
+        pub enum #message_enum_name {
+            #(#discriminator_names(#message_types),)*
+        }
+
+        #[derive(#input_derives)]
+        pub enum #input_enum_name {
+            #(#discriminator_names(#input_types),)*
+        }
+
+        impl near_async::messaging::CanSend<#message_enum_name> for #struct_name {
+            fn send(&self, message: #message_enum_name) {
+                match message {
+                    #(#message_enum_name::#discriminator_names(message) => self.#field_names.send(message),)*
+                }
+            }
+        }
+
+        #(impl From<#message_types> for #message_enum_name {
+            fn from(message: #message_types) -> Self {
+                #message_enum_name::#discriminator_names(message)
+            }
+        })*
+
+        impl #message_enum_name {
+            pub fn into_input(self) -> #input_enum_name {
+                match self {
+                    #(Self::#discriminator_names(msg) => #input_enum_name::#discriminator_names(#input_extractors),)*
+                }
+            }
+        }
+    }
+}
+
+#[cfg(test)]
+mod tests {
+    use quote::quote;
+
+    #[test]
+    fn test_derive_into_multi_send() {
+        let input = quote! {
+            struct TestSenders {
+                sender: Sender<String>,
+                async_sender: AsyncSender<String, u32>,
```

Comment:
nit: this pattern in test made me think for a bit that there always should be a pair of sender and async sender. May worth to break it somewhere.

Reply:
In practice there is only one of them in most cases. The reason why we need two of them is that sometimes we need a response and sometimes we don't. We usually don't need both cases - we either need a response or we don't.